### DISCUSSION TITLE
Local development

### DISCUSSION THREAD

**Discussion by SandraEliz** [open-sdg/open-sdg#1932] [Q&A] [answered]
Hello

I am carrying out the instructions for local development and it shows the following error:

![image](https://user-images.githubusercontent.com/117297044/215168008-bb2e718e-07c0-481f-8175-15ad8c6e3a71.png)

![image](https://user-images.githubusercontent.com/117297044/215168134-0880102a-05fa-4eba-adbd-cda7ae291add.png)

Can you advise me please?

Thanks

**Comment by brockfanning** [comment]
@SandraEliz Judging from the error message it is failing it [this point in the code](https://github.com/open-sdg/jekyll-open-sdg-plugins/blob/2.2.0-dev/lib/jekyll-open-sdg-plugins/fetch_remote_data.rb#L85). I'm just guessing because I can't see your code, but is it possible that you don't have "languages" in your site configuration?

**Reply by SandraEliz** [reply]
Hello

if i have that setup

![image](https://user-images.githubusercontent.com/117297044/215808727-7e9baa54-6119-4d2b-b0d8-043ca909b306.png)

**Comment by brockfanning** [comment]
@SandraEliz I can't recreate that when I run it locally. Have you changed any files in the repository?

**Comment by SandraEliz** [comment]
Hello

It already works, now my question is, when making changes to the files they are not reflected in the URL, what should be done to make these changes visible?

Thanks

**Reply by brockfanning** [reply]
@SandraEliz Can you give an example of a file you are changing? And also what is the URL you are looking at?

**Reply by SandraEliz** [reply]
Hello

Running bundle exec jekyll serve gives me a url to view the site.

I make changes to a data file, I change a number, and when I check the URL it is not reflected.

Am I missing something to do?

**Reply by SandraEliz** [reply]
You are also checking the following:

![image](https://user-images.githubusercontent.com/117297044/216083121-b799f623-09e8-4cb6-95e1-f27f0f2b9f0c.png)

**Reply by brockfanning** [reply]
@SandraEliz I assume that you have your site and data in two separate repositories - is that correct? If that is correct, please read on:

When you run `bundle exec jekyll serve` you will see a local site that is pulling data from whatever is in the site configuration as `remote_data_prefix`. For example, if your `remote_data_prefix` is `https://prep-odsmexico.github.io/data/` then this means that your local site will be pulling data from `https://prep-odsmexico.github.io/data/`.

It sounds like you are changing numbers in your local clone of the data repository. If that is correct, you might guess why this doesn't affect your local site: if your local site is pulling data from `https://prep-odsmexico.github.io/data/`, then any changes you make locally will *not* be reflected in the site.

If you want to change numbers locally and see them reflected in your local site, you would need to build the data locally and then change your site configuration to pull data from there. Our documentation at https://open-sdg.readthedocs.io/en/latest/development/ is incomplete because it does not cover this. So thank you for bringing this up!

Here is a rough draft of the steps. These assume that you have Python 3 installed.

1. Clone your data repository locally, and change any numbers you would like to change
2. In your local data repository, run: `pip install -r scripts/requirements.txt`
3. In your local data repository, run `python scripts/build_data.py`
4. In your local data repository, run `python -m http.server --directory _site`
5. In your local site repository, change `remote_data_prefix` to `http://localhost:8000`
6. In your local site repository, run `bundle exec jekyll serve`

Now you should see the changed numbers reflected locally.

The warning you posted about SSL certificates and rubygems.org, is unrelated to what I wrote above. That is probably related to the way in which you have installed Ruby. In case it is helpful, personally when I am using Windows I use WSL2 and I have Ruby installed using RVM.

**Comment by SandraEliz** [comment]
Hello

I already did the steps that you comment but it gives this error

![image](https://user-images.githubusercontent.com/117297044/216378498-9e3a5b7f-aa5c-494d-97af-f8f12192cba7.png)

**Reply by SandraEliz** [reply]
Hello

It no longer shows me that error, now it shows this:

![image](https://user-images.githubusercontent.com/117297044/216685722-f9480378-fd61-48e3-a6b0-33c7763a90b5.png)

**Reply by brockfanning** [reply]
@SandraEliz We need to figure out what URL you're local data is available at. You should able to reach it in a browser. For example do you see anything if you go to http://localhost:8000 in a browser? If not, when you ran the `python -m http.server --directory _site` command, what output did you see?

**Reply by SandraEliz** [reply]
![image](https://user-images.githubusercontent.com/117297044/216701732-b9d2f082-68d2-403f-99d7-29783ba967e5.png)

**Reply by brockfanning** [reply]
And does http://localhost:8000 work in a browser?

**Reply by SandraEliz** [reply]
![image](https://user-images.githubusercontent.com/117297044/216702837-ed930fc3-a4f6-48a8-9dd0-129d9e1d66c6.png)

**Reply by brockfanning** [reply]
@SandraEliz The challenge will be to figure out what the URL of the local data build is. I'm not sure how to help with that, since it depends probably on your operating system and Python environment. I can say that another alternative to `python -m http.server --directory _site` is to use a Node package called "live-server". I've also tried that in the past. To use that, the commands would be:

```
npm install --global live-server
cd _site
live-server --cors
```

**Reply by SandraEliz** [reply]
Hello

With the live-server --cors I can see the site, but it doesn't reflect the changes.

I already changed the remote_data_prefix for the url it shows, in this case it is http://127.0.0.1:8080

How can I get the site to regenerate when I make changes to the data files?

Thank you

**Reply by brockfanning** [reply, marked as answer]
@SandraEliz Ok, it is my understanding that you have everything working locally, except that you would like it to be automated such that:
1. You make a local change in the data repository
2. The data is automatically rebuilt
3. The site is automatically rebuilt

I suspect something like that is possible with some creative use of command-line tools, but I do not know how to do it.

I also suspect that would be much easier if the site and the data were in the same repository, instead of being in two separate repositories. If you are interested in trying that out, you could investigate the [open-sdg-simple-starter repository](https://github.com/open-sdg/open-sdg-simple-starter).

**Reply by brockfanning** [reply]
@SandraEliz Also re-reading your comment, maybe you are just looking for a quick way to rebuild the site? You can press "CTRL-c" to stop the site, and then run "bundle exec jekyll serve" again. This will rebuild the site.
